### PR TITLE
Eslint rewrite code only

### PR DESCRIPTION
**Pull request by andreaTP**
Here I can review :-)

### CODE REVIEW

File: src/allFiles.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { walk } from "./promisified"
```
Comment (andreaTP) [inline src/allFiles.ts]: you already have a file `fileUtils` doesn't this function belong to it?

File: src/configCreator.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import { cloneDeep, isEmpty, partition, fromPairs } from "lodash"
+import { allFilesNames } from "./allFiles"
+
+function patternsToRules(
+  patterns: Pattern[]
+): { [name: string]: Linter.RuleLevel | Linter.RuleLevelAndOptions } {
+  const pairs = patterns.map(pattern => {
+    const patternId = patternIdToEslint(pattern.patternId)
+    if (pattern.parameters) {
+      const [unnamedParameters, namedParameters] = partition(
```
Comment (andreaTP) [inline src/configCreator.ts]: you are doing a bunch of operations on `namedParameters` does it make sense to refactor in a class?

File: src/configCreator.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  srcDirPath: string,
+  codacyInput?: Codacyrc
+): Promise<CLIEngine.Options> {
+  if (codacyInput && codacyInput.tools) {
+    const eslintTool = codacyInput.tools.find(tool => tool.name === toolName)
+    if (eslintTool && eslintTool.patterns) {
+      const patterns = eslintTool.patterns
+      const result = cloneDeep(defaultOptions)
+      if (result.baseConfig) {
+        result.baseConfig.extends = [] // TODO: Maintain base configurations without rules
```
Comment (andreaTP) [inline src/configCreator.ts]: can you do the todo?

File: src/configCreator.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        result.baseConfig.overrides.extends = []
+        result.baseConfig.rules = patternsToRules(patterns)
+      }
+      result.useEslintrc = false
+      return result
+    }
+  }
+  const fileNames = await allFilesNames(srcDirPath)
+  if (fileNames.find(name => name.startsWith(".eslintrc")) !== undefined)
```
Comment (andreaTP) [inline src/configCreator.ts]: here I think you can get a wrong `.eslintrc` 
e.g. from a subfolder that can be a dead file in the repo

File: src/configCreator.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    return {}
+  else return defaultOptions
+}
+
+export async function configCreator(
+  srcDirPath: string,
+  codacyInput?: Codacyrc
+): Promise<[CLIEngine.Options, string[]?]> {
+  const options = createOptions(srcDirPath, codacyInput)
+  const files = codacyInput && codacyInput.files ? codacyInput.files : undefined
```
Comment (andreaTP) [inline src/configCreator.ts]: use an empty array instead of `undefined`

File: src/convertResults.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import { CLIEngine, Linter } from "eslint"
+import { CodacyResult } from "./model/CodacyResult"
+import { flatMap } from "lodash"
+import { patternIdToCodacy } from "./model/Patterns"
+
+export function convertResults(report: CLIEngine.LintReport): CodacyResult[] {
+  return flatMap(report.results, result => {
+    const filename = result.filePath
+    const pairs = result.messages.map(m => [m.ruleId, m]).filter(t => t[0]) as [
```
Comment (andreaTP) [inline src/convertResults.ts]: is this operation safe? can you explain the rationale behind the `[0]` ?

File: src/docGenerator.ts [hunk 7]
```diff
@@ -0,0 +1,146 @@
+import fs from "fs"
```
Comment (andreaTP) [inline src/docGenerator.ts]: in this file there are too many concerns
file system operations, json handling and more,
please split the different functionalities for clarity

File: src/docGenerator.ts [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+  Patterns,
+  PatternsEntry,
+  PatternsParameter
+} from "./model/Patterns"
+import { toolName, toolVersion } from "./toolMetadata"
+
+const writeFile = promisify(fs.writeFile)
+
+export function generatePatterns(): Patterns {
+  const rules = defaultEngine.getRules()
```
Comment (andreaTP) [inline src/docGenerator.ts]: shouldn't `rules` come as an argument to this function?

File: src/docGenerator.ts [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+  const patterns = Array.from(rules.entries()).map(
+    ([patternId, ruleModule]) => {
+      const meta = ruleModule && ruleModule.meta ? ruleModule.meta : undefined
+      const eslintCategory = meta && meta.docs ? meta.docs.category : undefined
+      const level: Level = fromEslintCategoryToLevel(eslintCategory)
+      const category: Category = patternId.includes("security")
```
Comment (andreaTP) [inline src/docGenerator.ts]: are we flattening the security subcategories?

File: src/docGenerator.ts [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        ? "Security"
+        : fromEslintCategoryToCategory(eslintCategory)
+      const parameters =
+        meta && meta.schema
+          ? fromEslintSchemaToParameters(meta.schema)
+          : undefined
```
Comment (andreaTP) [inline src/docGenerator.ts]: use an empty array instead of `undefined`

File: src/docGenerator.ts [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+      return new PatternsEntry(
+        patternIdToCodacy(patternId),
+        level,
+        category,
+        parameters && parameters.length > 0 ? parameters : undefined
```
Comment (andreaTP) [inline src/docGenerator.ts]: use an empty array instead of `undefined`

File: src/docGenerator.ts [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+      )
+    }
+  )
+  const entries = patterns.filter(x => x !== null) as PatternsEntry[]
```
Comment (andreaTP) [inline src/docGenerator.ts]: can you use fold or flatMap in first place, pretty sure you have something  alike in lodash

File: src/docGenerator.ts [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+  const rules = defaultEngine.getRules()
+  const descriptionEntries = Array.from(rules.entries()).map(
+    ([patternId, ruleModule]) => {
+      const eslintDescription =
+        ruleModule && ruleModule.meta && ruleModule.meta.docs
+          ? ruleModule.meta.docs.description
+          : undefined
+      const capitalize = (s: string) => s.charAt(0).toUpperCase() + s.slice(1)
+      const description = eslintDescription ? capitalize(eslintDescription) : ""
+      const title = patternId
```
Comment (andreaTP) [inline src/docGenerator.ts]: this needs for sure proper testing

Comment (andreaTP) [reply src/docGenerator.ts]: and should be extracted to his function

File: src/docGenerator.ts [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+type WithDefault = { default: any }
+
+function anyOfToArray(schema: JSONSchema4) {
+  return schema.anyOf ? schema.anyOf : [schema]
+}
+
+function fromEslintSchemaToParameters(
+  schema: JSONSchema4 | JSONSchema4[]
```
Comment (andreaTP) [inline src/docGenerator.ts]: just `JSONSchema4[]` please

File: src/docGenerator.ts [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+function eslintPatternIds(): Array<string> {
+  const rules = defaultEngine.getRules()
+  return Array.from(rules.keys()).filter(e => !e.includes("/"))
+}
+
+export function downloadDocs(
+  urlFromPatternId: (patternId: string) => string,
+  prefix: string | undefined = undefined
```
Comment (andreaTP) [inline src/docGenerator.ts]: empty string instead of `undefined` ?

File: src/model/CodacyInput.ts [hunk 16]
```diff
@@ -0,0 +1,41 @@
+export type ParameterValue = any
```
Comment (andreaTP) [inline src/model/CodacyInput.ts]: argh :-(

File: src/model/CodacyInput.ts [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    this.name = name
+    this.value = value
+  }
+}
+
+export class Pattern {
+  readonly patternId: string
+  readonly parameters?: Parameter[]
+
+  constructor(patternId: string, parameters?: Parameter[]) {
```
Comment (andreaTP) [inline src/model/CodacyInput.ts]: arrays should degraded to empty array not to `undefined`

Comment (andreaTP) [reply src/model/CodacyInput.ts]: everywhere in this file

File: src/model/Description.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+    this.description = description
+  }
+}
+
+export class DescriptionEntry {
+  readonly patternId: string
+  readonly title: string
+  readonly description: string
+  readonly timeToFix: number
+  readonly parameters?: DescriptionParameter[]
```
Comment (andreaTP) [inline src/model/Description.ts]: not optional

File: src/model/Patterns.ts [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+      return "Info"
+    case "Removed":
+      return "Info"
+    default:
+      return "Info"
+  }
+}
+
+export function patternIdToCodacy(patternId: string): string {
+  return patternId.split("/").join("_")
```
Comment (andreaTP) [inline src/model/Patterns.ts]: use regex instead

File: src/model/Patterns.ts [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+export function patternIdToEslint(patternId: string): string {
+  return patternId.split("_").join("/")
```
Comment (andreaTP) [inline src/model/Patterns.ts]: use regex instead

Comment (andreaTP) [reply src/model/Patterns.ts]: and this is not safe ...

File: src/parseTimeoutSeconds.ts [hunk 21]
```diff
@@ -0,0 +1,26 @@
+export const defaultTimeout = 15 * 60
+
+export function parseTimeoutSeconds(timeoutString?: string): number {
```
Comment (andreaTP) [inline src/parseTimeoutSeconds.ts]: there should be a library to do this for you, instead of running your own implementation

File: tsconfig.tsbuildinfo [hunk 22]
```diff
@@ -0,0 +1,2328 @@
+{
```
Comment (andreaTP) [inline tsconfig.tsbuildinfo]: is this supposed to be committed?

**Review by andreaTP — COMMENTED** [review]
Ok, I started with a bunch of comments, in general looks nice but we need a few more unit tests where logic is involved.
cc. @lolgab

File: src/convertResults.ts (hunk 6 above)
Comment (lolgab) [reply src/convertResults.ts]: I map first the messages to a [tuple](https://www.typescriptlang.org/docs/handbook/basic-types.html#tuple).
Then I filter them by `ruleId` (to check that ruleId is defined, since it is `string | undefined`).
It is not checked by the compiler, so you can put whatever number you want.
I'm changing it to `.filter(([ruleId, _m]) => ruleId)`. Yet not checked at compile time (you can put whatever number of elements in the tuple matching) but, at least, more readable

File: src/docGenerator.ts (hunk 14 above)
Comment (lolgab) [reply src/docGenerator.ts]: Eslint can manage vanilla `JSONSchema4`s and arrays of `JSONSchema4`. And they have different structures when you use the array form or the non array one.
[Source](https://eslint.org/docs/developer-guide/working-with-rules#options-schemas)

File: src/model/CodacyInput.ts (hunk 16 above)
Comment (lolgab) [reply src/model/CodacyInput.ts]: Can be anything representable as Json value.. That's why `any`

File: src/model/CodacyInput.ts (hunk 17 above)
Comment (lolgab) [reply src/model/CodacyInput.ts]: I used these types because it matches the structure of the files we use in Codacy.
For example, a valid Pattern in our .codacyrc file can have no parameters section at all.
So I can parse the Json file and have a Pattern object directly.
Otherwise I should define a structure like this for parsing, and another structure with no undefined fields and convert by hand to it after parsing.

File: src/model/Description.ts (hunk 18 above)
Comment (lolgab) [reply src/model/Description.ts]: Same story as above, facility for parsing.

File: src/model/Patterns.ts (hunk 20 above)
Comment (lolgab) [reply src/model/Patterns.ts]: I know it's not safe.. But:
1) couldn't find rules with underscore names.
2) did this for retro-compatibility with the old tool, so old patterns' settings are maintained.

File: src/parseTimeoutSeconds.ts (hunk 21 above)
Comment (lolgab) [reply src/parseTimeoutSeconds.ts]: I couldn't find anything that matches the Scala's timeout format we use

File: src/promisified.ts [hunk 23]
```diff
@@ -0,0 +1,7 @@
+import fs from "fs"
```
[the remark this answers is not in the document]
Comment (lolgab) [reply src/promisified.ts]: Done ✅

File: src/docGenerator.ts (hunk 15 above)
Comment (lolgab) [reply src/docGenerator.ts]: Done ✅

File: .vscode/settings.json [hunk 24]
```diff
@@ -0,0 +1,3 @@
+{
```
[the remark this answers is not in the document]
Comment (lolgab) [reply .vscode/settings.json]: Done ✅

File: src/allFiles.ts (hunk 1 above)
Comment (lolgab) [reply src/allFiles.ts]: Moved to `fileUtils.ts` ✅

File: src/configCreator.ts (hunk 3 above)
Comment (lolgab) [reply src/configCreator.ts]: Wise question :-D
Sadly this is not possible with EsLint, you can't extend a configuration without inheriting its rules.
I forgot to remove the TODO comment.

File: src/configCreator.ts (hunk 4 above)
Comment (lolgab) [reply src/configCreator.ts]: Here the problem is tricky.
Because Eslint support `.eslintrc` files in any subfolder.
Do you have suggestions here?

File: src/model/Patterns.ts (hunk 19 above)
Comment (lolgab) [reply src/model/Patterns.ts]: Done ✅

File: src/model/Patterns.ts (hunk 20 above)
Comment (lolgab) [reply src/model/Patterns.ts]: ✅ Moved to regex, added test too.

File: src/configCreator.ts (hunk 5 above)
Comment (lolgab) [reply src/configCreator.ts]: Done ✅

File: src/docGenerator.ts (hunk 10 above)
Comment (lolgab) [reply src/docGenerator.ts]: The same here, I want to create a json object without the parameters field, not one with a parameters field with an empty array in the `patterns.json`

File: src/docGenerator.ts (hunk 13 above)
Comment (lolgab) [reply src/docGenerator.ts]: Done ✅

File: src/configCreator.ts (hunk 4 above)
Comment (rtfpessoa) [reply src/configCreator.ts]: You cannot search for the file. ESLint needs to find the files themselves when the user wants config file.
It needs to support multiples files in the tree, you don't want to do that. Please read the old code before doing this to make sure you cover all the cases so we don't introduce bugs again.

**Review by rtfpessoa — CHANGES_REQUESTED** [review]

**Comment by lolgab** [comment]
The thing here is that I check that only to see if the user has actually a `.eslintrc` config in the repository.. If the repository contains that file, I want to have an empty baseconfig so eslint can pick only the one of the user.. I don't want this to happen again: https://github.com/codacy/codacy-meta/issues/80

**Comment by rtfpessoa** [comment]
That seems unrelated. The only error I see there is probably them having a
different version of the configs.
And I cannot ser advantage to have anything besides what we have when we
either run like the user or we disable configs and pass the one generated
from ui
-- 




Rodrigo Fernandes

Software Engineer, Codacy


e: rodrigo@codacy.com

w: https://www.codacy.com

New York | Lisbon

**Comment by lolgab** [comment]
Ok, so I removed that code and I give and empty config when there are no patterns defined in Codacy.
Regardless of a .eslintrc being there

File: src/configCreator.ts (hunk 2 above)
Comment (lolgab) [reply src/configCreator.ts]: I moved them to a separate file and I added a test for them.

File: package.json [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+    "type": "git",
+    "url": "git+https://github.com/codacy/codacy-eslint.git"
+  },
+  "license": "Apache-2.0",
+  "main": "dist/src/index.js",
+  "directories": {
+    "src": "./src",
+    "test": "./src/test"
+  },
+  "scripts": {
```
[the remark this answers is not in the document]
Comment (lolgab) [reply package.json]: I'm running directly from node. Since it requires files in special places "/src" directory "/.codacyrc" file, I don't see value in running it outside of the docker container.

File: src/docGenerator.ts (hunk 8 above)
Comment (lolgab) [reply src/docGenerator.ts]: Done ✅ 
I made rules field of the DocGenerator class

File: src/docGenerator.ts (hunk 9 above)
Comment (lolgab) [reply src/docGenerator.ts]: Yes, I don't see an easy way to filter out the security subcategories.
I would leave this to a subsequent PR, if you are ok with it.

File: src/docGenerator.ts (hunk 11 above)
Comment (lolgab) [reply src/docGenerator.ts]: Same as before, it has different semantics and in our format we have no "parameters" defined, which maps 1:1 to the JSON format.

File: src/docGenerator.ts (hunk 12 above)
Comment (lolgab) [reply src/docGenerator.ts]: ✅Done

File: tsconfig.tsbuildinfo (hunk 22 above)
Comment (lolgab) [reply tsconfig.tsbuildinfo]: No, it was already .gitignored 😆
Removed ✅

**Comment by rtfpessoa** [comment]
@lolgab wait a minute. So that was the default case?
I did not understand that before sorry :disappointed: 

In that case, I guess ESlint before did not have that case yet.
Do you have a way to do that search using ESLint code? If not I guess that code should be fine.

File: package.json (hunk 25 above)
[the remark this answers is not in the document]
Comment (andreaTP) [reply package.json]: :+1:

File: src/docGenerator.ts [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+          description,
+          timeToFix
+        )
+      }
+    )
+    return descriptionEntries.filter(x => x !== null) as DescriptionEntry[]
+  }
+
+  private fromEslintSchemaToParameters(
+    schema: JSONSchema4 | JSONSchema4[]
```
Comment (andreaTP) [inline src/docGenerator.ts]: can this method take only `JSONSchema4[]` ?

File: src/docGenerator.ts [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+  ): PatternsParameter[] {
+    const anyOfToArray = (schema: JSONSchema4) => {
+      return schema.anyOf ? schema.anyOf : [schema]
```
Comment (andreaTP) [inline src/docGenerator.ts]: I don't like the position of this `return`
and is quite hard to follow all of the exit conditions of this method

File: src/docGenerator.ts [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+    const filteredPatternIds = patternIds.filter(patternId =>
+      patternId.startsWith(longPrefix)
+    )
+    return filteredPatternIds.map(patternId =>
+      patternId.substring(longPrefix.length)
+    )
+  }
+
+  private eslintPatternIds(): Array<string> {
+    return Array.from(this.rules.keys()).filter(e => !e.includes("/"))
```
Comment (andreaTP) [inline src/docGenerator.ts]: this line deserves at least a comment explaining the reasoning behind

File: src/docGenerator.ts [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+      const result = await fetch(url)
+      if (result.ok) {
+        const text = await result.text()
+        const filename =
+          "docs/description/" +
+          (prefix.length > 0 ? prefix + "_" : "") +
+          patternIdToCodacy(pattern) +
+          ".md"
+        return writeFile(filename, text)
+      } else return Promise.resolve()
```
Comment (andreaTP) [inline src/docGenerator.ts]: why we are swallowing errors?
`Promise.reject` instead

File: src/model/CodacyInput.ts (hunk 16 above)
Comment (andreaTP) [reply src/model/CodacyInput.ts]: argh:
https://github.com/microsoft/TypeScript/issues/1897
it's not yet encoded in std lib :-(

File: src/parseTimeoutSeconds.ts (hunk 21 above)
Comment (andreaTP) [reply src/parseTimeoutSeconds.ts]: this:
https://github.com/josephtzeng/hocon-parser/blob/3baba010f08e4a25d08ca190053693d30ed74dc1/tests/resources/configs/include-from-list/test01.conf#L58
looks appropriate, isn't it?

File: src/test/fileUtils.spec.ts [hunk 30]
```diff
@@ -0,0 +1,48 @@
+import { parseCodacyrcFile } from "../fileUtils"
+import { deepEqual } from "assert"
+import { Codacyrc } from "../model/CodacyInput"
+
+describe("main", () => {
+  it("should parse a codacyrc file", () => {
```
Comment (andreaTP) [inline src/test/fileUtils.spec.ts]: please have negative tests for checking the behavior with invalid `codacyrc` files

File: src/test/timeout.spec.ts [hunk 31]
```diff
@@ -0,0 +1,26 @@
+import { equal } from "assert"
+import { parseTimeoutSeconds, defaultTimeout } from "../parseTimeoutSeconds"
+
+describe("parseTimeoutSeconds", () => {
```
Comment (andreaTP) [inline src/test/timeout.spec.ts]: in case we have to keep our implementation, here is the reference in shocon:
https://github.com/akka-js/shocon/blob/master/facade/shared/src/main/scala/com/typesafe/config/Config.scala#L197

**Review by andreaTP — COMMENTED** [review]
going in the right direction!!!

File: src/docGenerator.ts (hunk 29 above)
Comment (lolgab) [reply src/docGenerator.ts]: ✅Done (found a bug too, thank you 🙏)

File: src/test/fileUtils.spec.ts (hunk 30 above)
Comment (lolgab) [reply src/test/fileUtils.spec.ts]: ✅Done

File: src/docGenerator.ts (hunk 28 above)
Comment (lolgab) [reply src/docGenerator.ts]: ✅Done

File: src/docGenerator.ts (hunk 26 above)
Comment (lolgab) [reply src/docGenerator.ts]: Already answered [here](https://github.com/codacy/codacy-eslint-strawman/pull/2#discussion_r357563186)
Github hides a lot of comments, maybe you are not seeing them (there is a button "Show hidden conversations")

File: src/docGenerator.ts (hunk 27 above)
Comment (lolgab) [reply src/docGenerator.ts]: ✅Fixed

**Comment by lolgab** [comment]
I don't know why github doesn't let me answer on this:
```
this:
https://github.com/josephtzeng/hocon-parser/blob/3baba010f08e4a25d08ca190053693d30ed74dc1/tests/resources/configs/include-from-list/test01.conf#L58
looks appropriate, isn't it?
```
By the way it has no typescript facades :(

File: src/test/timeout.spec.ts (hunk 31 above)
Comment (lolgab) [reply src/test/timeout.spec.ts]: I remember we used only the outputs of `scala.concurrent.duration`'s`toString` with seconds, minutes and hours, not the entire HOCON standard. The same we do in many tools (pylint, prospector, codacy-sonar-csharp, etc.)
cc. @DReigada

File: src/parseTimeoutSeconds.ts (hunk 21 above)
Comment (lolgab) [reply src/parseTimeoutSeconds.ts]: it has no typescript facades :(

File: src/test/timeout.spec.ts (hunk 31 above)
Comment (DReigada) [reply src/test/timeout.spec.ts]: This was a mistake on our side, the timeout should be set as a universal format (e.g. number of seconds) and not scala duration. We should just ignore the timeouts in tools that do not use the scala seed. And eventually in the future we will add a new env var which passes the timeout correctly

Comment (lolgab) [reply src/test/timeout.spec.ts]: I would say to leave this implementation until the PR to introduce the new `TIMEOUT_SECONDS` env variable is merged in codacy internals.. Then I will remove the timeout parsing implementation and switch to the new one.

Comment (andreaTP) [reply src/test/timeout.spec.ts]: ok with this
